ext/semian_cb_data/semian_cb_data.c

```diff
@@ -0,0 +1,900 @@
+static const int kCBSemaphoreCount = 1; // # semaphores to be acquired
+static const int kCBTicketMax = 1;
+static const int kCBInitializeWaitTimeout = 5; /* seconds */
+static const int kCBIndexTicketLock = 0;
+static const int kCBInternalTimeout = 5; /* seconds */
+
+static struct sembuf decrement; // = { kCBIndexTicketLock, -1, SEM_UNDO};
+static struct sembuf increment; // = { kCBIndexTicketLock, 1, SEM_UNDO};
+
+static VALUE eInternal, eSyscall, eTimeout; // Semian errors
+
+static void semian_cb_data_mark(void *ptr);
+static void semian_cb_data_free(void *ptr);
+static size_t semian_cb_data_memsize(const void *ptr);
+static VALUE semian_cb_data_alloc(VALUE klass);
+static VALUE semian_cb_data_init(VALUE self, VALUE name, VALUE size, VALUE permissions);
+static VALUE semian_cb_data_clean(VALUE self);
+static void set_semaphore_permissions(int sem_id, int permissions);
+static void configure_tickets(int sem_id, int tickets, int should_initialize);
+static int create_semaphore(int key, int permissions, int *created);
+static VALUE semian_cb_data_acquire_semaphore (VALUE self, VALUE permissions);
+static VALUE semian_cb_data_delete_semaphore(VALUE self);
+static VALUE semian_cb_data_lock(VALUE self);
+static VALUE semian_cb_data_unlock(VALUE self);
+static void *semian_cb_data_lock_without_gvl(void *self);
+static void *semian_cb_data_unlock_without_gvl(void *self);
+static VALUE semian_cb_data_acquire_memory(VALUE self, VALUE permissions);
+static void semian_cb_data_delete_memory_inner (semian_cb_data *ptr);
+static VALUE semian_cb_data_delete_memory (VALUE self);
+static VALUE semian_cb_data_get_successes(VALUE self);
+static VALUE semian_cb_data_set_successes(VALUE self, VALUE num);
+static VALUE semian_cb_data_semid(VALUE self);
+static VALUE semian_cb_data_shmid(VALUE self);
+static VALUE semian_cb_data_array_at_index(VALUE self, VALUE idx);
+static VALUE semian_cb_data_array_set_index(VALUE self, VALUE idx, VALUE val);
+static VALUE semian_cb_data_array_length(VALUE self);
+static VALUE semian_cb_data_set_push_back(VALUE self, VALUE num);
+static VALUE semian_cb_data_set_pop_back(VALUE self);
+static VALUE semian_cb_data_set_push_front(VALUE self, VALUE num);
+static VALUE semian_cb_data_set_pop_front(VALUE self);
+
+// needed for TypedData_Make_Struct && TypedData_Get_Struct
+static const rb_data_type_t
+semian_cb_data_type = {
+  "semian_cb_data",
+  {
+    semian_cb_data_mark,
+    semian_cb_data_free,
+    semian_cb_data_memsize
+  },
+  NULL, NULL, RUBY_TYPED_FREE_IMMEDIATELY
+};
+
+/*
+ * Generate key
+ */
+static key_t
+generate_key(const char *name)
+{
+  union {
+    unsigned char str[SHA_DIGEST_LENGTH];
+    key_t key;
+  } digest;
+  SHA1((const unsigned char *) name, strlen(name), digest.str);
+  /* TODO: compile-time assertion that sizeof(key_t) > SHA_DIGEST_LENGTH */
+  return digest.key;
+}
+
+/*
+ * Log errors
+ */
+static void
+raise_semian_syscall_error(const char *syscall, int error_num)
```

Comment:
How would the #include work? I'm imagining a common semian_main.c+h file or something that both semian.c and the circuit breaker file include.